timelock-impl/src/main/java/com/palantir/atlasdb/backup/AtlasBackupResource.java

```diff
@@ -0,0 +1,178 @@
+        return new JerseyAtlasBackupServiceAdapter(new AtlasBackupResource(redirectRetryTargeter, timelockServices));
+    }
+
+    @Override
+    public ListenableFuture<PrepareBackupResponse> prepareBackup(AuthHeader authHeader, PrepareBackupRequest request) {
+        return handleExceptions(() -> Futures.immediateFuture(prepareBackupInternal(request)));
+    }
+
+    private PrepareBackupResponse prepareBackupInternal(PrepareBackupRequest request) {
+        Set<BackupToken> preparedBackups = request.getNamespaces().stream()
+                .map(this::prepareBackup)
+                .flatMap(Optional::stream)
+                .collect(Collectors.toSet());
+        return PrepareBackupResponse.of(preparedBackups);
+    }
+
+    Optional<BackupToken> prepareBackup(Namespace namespace) {
+        try {
+            return Optional.of(tryPrepareBackup(namespace));
+        } catch (Exception ex) {
+            log.info("Failed to prepare backup for namespace", SafeArg.of("namespace", namespace), ex);
+            return Optional.empty();
+        }
+    }
+
+    private BackupToken tryPrepareBackup(Namespace namespace) {
+        AsyncTimelockService timelock = timelock(namespace);
+        LockImmutableTimestampResponse response = timelock.lockImmutableTimestamp(IdentifiedTimeLockRequest.create());
+        long timestamp = timelock.getFreshTimestamp();
+        return BackupToken.builder()
+                .namespace(namespace)
+                .lockToken(response.getLock())
+                .immutableTimestamp(response.getImmutableTimestamp())
+                .backupStartTimestamp(timestamp)
+                .build();
+    }
+
+    @Override
+    public ListenableFuture<CompleteBackupResponse> completeBackup(
+            AuthHeader authHeader, CompleteBackupRequest request) {
+        Map<BackupToken, ListenableFuture<Optional<BackupToken>>> futureMap =
+                request.getBackupTokens().stream().collect(Collectors.toMap(token -> token, this::completeBackupAsync));
+        ListenableFuture<Map<BackupToken, BackupToken>> singleFuture = AtlasFutures.futuresCombiner(
+                        MoreExecutors.newDirectExecutorService())
+                .allAsMap(futureMap);
+        return Futures.transform(
+                singleFuture,
+                map -> CompleteBackupResponse.of(new HashSet<>(map.values())),
+                MoreExecutors.directExecutor());
+    }
+
+    private ListenableFuture<Optional<BackupToken>> completeBackupAsync(BackupToken backupToken) {
+        return Futures.transform(
+                unlockAsync(backupToken),
+                tokens -> fetchFastForwardTimestampIfSuccessful(backupToken, tokens),
+                MoreExecutors.directExecutor());
+    }
+
+    private ListenableFuture<Set<LockToken>> unlockAsync(BackupToken backupToken) {
+        return timelock(backupToken.getNamespace()).unlock(Set.of(backupToken.getLockToken()));
+    }
```

Comment:
This may simplify code:
```suggestion
    @SuppressWarnings("ConstantConditions") // Set of locks is never null
    private ListenableFuture<Optional<LockToken>> maybeUnlock(BackupToken backupToken) {
        return Futures.transform(timelock(backupToken.getNamespace()).unlock(Set.of(backupToken.getLockToken())),
                set -> set.stream().findFirst(),
                MoreExecutors.directExecutor());
    }
```